Document --backup import behaviour (#1788)
iterative/dvc.org: content/docs/command-reference/import.md

they won't be updated. Use `dvc update` to update the downloaded data artifact
from the source repo.

### Backup mode

To avoid duplicating large files, `dvc import` does not push imported files to

> any [remotes](/doc/command-reference/remote) by default. This means that

Reviewer: ```suggestion
[remote storage](/doc/command-reference/remote) by default. This means that
```

> any [remotes](/doc/command-reference/remote) by default. This means that when
> `dvc pull` or `dvc fetch` is run, DVC retrieves imports from their source. So if

Reviewer: ```suggestion
any [remotes](/doc/command-reference/remote) by default. This means that
`dvc pull`, `dvc fetch`, and `dvc repro` retrieve imports from their source. So if
```

> the original source of an import is deleted, then the files are not backed up in
> your remote, and data may be lost.

Reviewer: ```suggestion
the original source of an import is deleted, then the files are not backed up
anywhere, and data may be lost.
```

To prevent this, you can use the `--backup` option to request DVC to backup the

> data imports to a remote. This also means that `dvc pull` or `dvc fetch`

Reviewer: ```suggestion
data imports to a remote. This also means that `dvc pull/fetch/repro`
```

> data imports to remote storage. This also means that `dvc pull` or `dvc fetch`

Reviewer: > `dvc import` does not push imported files

Will `import` do the pushing when using --backup? Or that just tells `push` to do it when/if run? I'm guessing the latter but it's not clear from the text above (also in the 2nd paragraph). Feel free to update it or just lmk what the answer is and I will 👍

Author: Yes, the latter: I'll tweak

> data imports to remote storage. This also means that `dvc pull` or `dvc fetch`

Reviewer: ```suggestion
data imports to a remote. This also means that `dvc pull` or `dvc fetch`
```

no longer need to access import sources, so these imports can be restored from

> remote storage like typical data tracked with `dvc add`.

Reviewer: ```suggestion
the remote like regular data tracked with `dvc add`.
```

> Imports using _backup mode_ have a flag in their `.dvc` file `outs` section:
> `backup: true`. To revert to normal import behaviour, remove this flag or set it
> to `false`.

Reviewer: ```suggestion
Backup mode is enabled by the `backup: true` key/value pair set in the `outs`
section of the import `.dvc` file. To switch to regular import behavior, set its
value to `false` (or remove the field altogether).
```

they won't be updated. Use `dvc update` to update the downloaded data artifact
from the source repo.

### Backup mode

> To avoid duplicating large files, `dvc import` does not push imported files to
> [remote storage](/doc/command-reference/remote) by default. This means that

Reviewer: ```suggestion
To avoid duplicating large files, imported data is not pushed to
[remote storage](/doc/command-reference/remote) by default (see `dvc push`). This means that
```

> [remote storage](/doc/command-reference/remote) by default (see `dvc push`). This means that
> `dvc pull`, `dvc fetch`, and `dvc repro` retrieve imports from their source. So if

Reviewer: ```suggestion
[remote storage](/doc/command-reference/remote) by default (see `dvc push`).
This means that `dvc pull`, `dvc fetch`, and `dvc repro` retrieve imports from
their source (found in the `deps` field of the `.dvc` file). So if
```

their source (found in the `deps` field of the `.dvc` file). So if
the original source of an import is deleted, then the files are not backed up
anywhere, and data may be lost.

To prevent this, you can use the `--backup` option to mark the imports to be

> pushed to a DVC remote when `dvc push` is run.  This also means that
> `dvc pull/fetch/repro` no longer access import sources, so these imports
> can be restored from the remote like regular data tracked with `dvc add`.

Reviewer: ```suggestion
pushed to a DVC remote when `dvc push` is run. This also means that
`dvc pull/fetch/repro` no longer access import sources, going to the remote
instead, like with regular project data tracked with `dvc add`. Use
`dvc update` to get the (latest) data from its import source.
```

they won't be updated. Use `dvc update` to update the downloaded data artifact
from the source repo.

### Backup mode

To avoid duplicating large files, imported data is not pushed to
[remote storage](/doc/command-reference/remote) by default (see `dvc push`).
This means that `dvc pull`, `dvc fetch`, and `dvc repro` retrieve imports from
their source (found in the `deps` field of the `.dvc` file). So if
the original source of an import is deleted, then the files are not backed up
anywhere, and data may be lost.

> To prevent this, you can use the `--backup` option to mark the imports to be
> pushed to a DVC remote when `dvc push` is run. This also means that

Reviewer: ```suggestion
As an alternative, you can use the `--backup` option so the import is
pushed to a DVC remote when `dvc push` is run. This also means that
```

> pushed to the remote when `dvc push` is run. 
> This also means that `dvc pull/fetch/repro`

Reviewer: ```suggestion
pushed to a DVC remote when `dvc push` is run.  This also means that
`dvc pull/fetch/repro`
```

> `dvc pull/fetch/repro`
> no longer needs to access import sources, so these imports can be restored from
> the remote like regular data tracked with `dvc add`.

Reviewer: ```suggestion
`dvc pull/fetch/repro` no longer access import sources, so these imports can be restored from
the remote like regular data tracked with `dvc add`.
```

> instead, like with regular project data tracked with `dvc add`. Use
> `dvc update` to get the (latest) data from its import source.

Reviewer: ```suggestion
instead, like with regular project data tracked with `dvc add`. You can still
use `dvc update` to get the latest data from its import source.
```

Backup mode is enabled with a `backup: true` key/value pair in the `outs`
section of the import `.dvc` file. To switch to regular import behavior, set

> its value to `false` (or remove the field altogether).

Reviewer: ```suggestion
its value to `false`, or re-import the data without using the `--backup` option.
```

they won't be updated. Use `dvc update` to update the downloaded data artifact
from the source repo.

> ### Store mode

Reviewer: ```suggestion
### Enabling remote storage
```

To avoid duplicating large files, imported data is not pushed to
[remote storage](/doc/command-reference/remote) by default (see `dvc push`).
This means that `dvc pull`, `dvc fetch`, and `dvc repro` retrieve imports from
their source (found in the `deps` field of the `.dvc` file). So if the original
source of an import is deleted, then the files are not backed up anywhere, and
data may be lost.

As an alternative, you can use the `--store` option so the import is pushed to a
DVC remote when `dvc push` is run. This also means that `dvc pull/fetch/repro`
no longer access import sources, going to the remote instead, like with regular
project data tracked with `dvc add`. You can still use `dvc update` to get the

> latest data from its import source.

Reviewer: ```suggestion
latest import data from its source.
```

> Store mode is enabled with a `store: true` key/value pair in the `outs` section
> of the import `.dvc` file. To switch to regular import behavior, set its value
> to `false`, or re-import the data without using the `--store` option.

Reviewer: ```suggestion
Remote storage is enabled with a `store: true` key/value pair in the `outs`
section of the import `.dvc` file. To switch to regular import behavior, set its
value to `false`, or re-import the data without using the `--store` option.
```

> section of the import `.dvc` file. To switch to regular import behavior, set its
> value to `false`, or re-import the data without using the `--store` option.

Reviewer: ```suggestion
section of the import `.dvc` file. To switch to regular import behavior, change
its value to `false` manually, or re-import the data without using the `--store` =
option.
```

  > [Importing and updating fixed revisions](#example-importing-and-updating-fixed-revisions)
  > example below).

> - `-b`, `--backup` - imports files in "backup mode": see Backup Mode section for
>   details.

Reviewer: ```suggestion
- `-b`, `--backup` - turns _backup mode_ on for this import, so it can be pushed
  to remote storage (see [Backup mode](#backup-mode) for details).
```

- `-h`, `--help` - prints the usage/help message, and exit.

- `-q`, `--quiet` - do not write anything to standard output. Exit with 0 if no
  problems arise, otherwise 1.

- `-v`, `--verbose` - displays detailed tracing information.

> ## Backup Mode

Reviewer: ```suggestion
## Backup mode
```

To avoid duplicating large files, `dvc import` does not push imported files to

> any configured remotes for your project by default. This means that when

Reviewer: ```suggestion
any remotes for your project by default. This means that when
```

> `dvc pull/fetch` is run, DVC retrieves imports from their source.

Reviewer: > when `dvc pull/fetch` is run, DVC retrieves imports from their source.

I didn't know that. Are you sure? I think you have to `dvc repro`. Please try it out 🙂

Author: It does! That was the motivation for this change actually, since I have a lot of git imports and all the repos were getting cloned each time I ran `dvc pull`. As a test, I ran `dvc pull -v` in a repo with a single import in it, like this:

```
md5: ca4361da725ee43b1e01ffe75f96b42a
frozen: true
deps:
- path: file.txt
  repo:
    url: ../imp
    rev_lock: f7d942abf891b4a0c29f8dec9e6728870862b012
outs:
- md5: f701279bc9ac499afb1f556e22c23cb2
  path: file.txt
```

Even after the import and the cache were present and up to date, the output contains

```
...
2020-09-20 22:20:08,091 DEBUG: Collecting information from local cache...
2020-09-20 22:20:08,094 DEBUG: Creating external repo ../imp@f7d942abf891b4a0c29f8dec9e6728870862b012                                                                                
2020-09-20 22:20:08,095 DEBUG: erepo: git clone '../imp' to a temporary dir
2020-09-20 22:20:08,192 DEBUG: Saving '../../../../private/var/folders/s8/gg_nx8mn23sbqxbj1021tn6c0000gn/T/tmpbg9ac34ldvc-clone/file.txt' to '.dvc/cache/f7/01279bc9ac499afb1f556e22c23cb2'.
...
```

Reviewer: Interesting. I should know this 😅 but also `repro` will do it so we just need to mention that one too in the sentence, please.

I'll make a separate issue for us to mention that fetch and pull re-import data...

Reviewer: > Even after the import and the cache were present and up to date (the repo is cloned)

On this, did DVC actually re-download the data? OR just clone the repo to check what the latest version was? Because if the former case, it sound like a bug to report in the core repo (shouldn't re-download imports if you already have the latest version).

Author: I think just clone to check but, in the case of git imports, that's the same since you need to redownload the entire git repo to clone it.

Reviewer: Hmmm. I think it could/should just clone the bare repo and check the latest revision where the file was modified (blame) is <= the `rev` in the .dvc file (`deps.repo.rev` field, to be precise.) Anyway... Do you see any more -v output that indicates whether the file was actually pulled/checked-out from the Git remote? It may be hard to determine, actually. I can take a look later if not so don;t worry too much about this! 🙂 Thanks!

Reviewer: So yes, `dvc pull` does download and overwrite the existing cached file even if it's the latest version (sounds like a bug to me, opened iterative/dvc/issues/4597). But only for Git imports, I think.

Anyway, resolving here!

> `dvc pull/fetch` is run, DVC retrieves imports from their source instead of from
> your remote. It also means that if the source of the import is deleted then the

Reviewer: ```suggestion
`dvc pull/fetch` is run, DVC retrieves imports from their source.
So if the original source of an import is deleted, then the
```

files are not backed up in your remote, and data may be lost.

To prevent this, you can use the `--backup` option to request DVC to backup the

> data imports to your project's remote. This also means that `dvc pull/fetch` no

Reviewer: ```suggestion
data imports to remote storage. This also means that `dvc pull/fetch` no
```

> longer needs to access import sources, so interacts only with the DVC remote,
> and that backed-up imports can be restored just like files that were
> `dvc add`ed.

Reviewer: ```suggestion
longer needs to access import sources, so
these imports can be restored from remote storage like typical
data tracked with `dvc add`.
```

To update backed-up imports, use `dvc update` like usual.

> 

Reviewer: Let's put this whole section as an H3 inside the **Description**, just before **Options**, please.

Reviewer: - [ ] Also, it should probably briefly describe what happens in the .dvc file (is there a new field?) and how to remove backup mode.
- [ ] Info about this field should also be added to https://dvc.org/doc/user-guide/dvc-files-and-directories#dvc-files please.

Author: Done and done. I already had some text in content/docs/user-guide/dvc-files-and-directories.md, though not loads. Is that fine?

Reviewer: Great, thanks. I didn't see those changes in the other file, yes that's enough! I'll take it from there 🙂

> 
> To update backed-up imports, use `dvc update` like usual.

Reviewer: No new info, it seems 🙂 
```suggestion
```

  > [Importing and updating fixed revisions](#example-importing-and-updating-fixed-revisions)
  > example below).

- `-b`, `--backup` - turns [backup mode](#backup-mode) on for the import,

>   so it can be pushed to remote storage.

Reviewer: ```suggestion
  so it can be pushed to remote storage (in case the original source is lost).
```

> - `-b`, `--backup` - turns _backup mode_ on for this import, so it can be pushed
>   to remote storage (see [Backup mode](#backup-mode) for details).

Reviewer: ```suggestion
- `-b`, `--backup` - turns [backup mode](#backup-mode) on for the import,
  so it can be pushed to remote storage.
```

  > [Importing and updating fixed revisions](#example-importing-and-updating-fixed-revisions)
  > example below).

- `-b`, `--backup` - turns [backup mode](#backup-mode) on for the import,

>   so it can be pushed to remote storage (in case the original source is lost).

Reviewer: What about `import-url`, does it have `--backup`? Probably should?

Author: `import-url` always made files in "backup mode", and still does. I haven't provided any option to disable backup mode for it though. I could do, or alternatively its behaviour could be documented.

Reviewer: Ah yeah I was confused. No need to mention backup mode in import-url, as it's explained in the DVC metafiles doc.

  > [Importing and updating fixed revisions](#example-importing-and-updating-fixed-revisions)
  > example below).

- `-s`, `--store` - [enables remote storage](#enabling-remote-storage) for the

>   import, so it can be pushed and pulled like regular project data (e.g. in

Reviewer: Maybe link pushed and pulled to the cmd refs?

Reviewer: Yeah
```suggestion
  import, so it can be [pushed](/doc/command-reference/push) and 
  [pulled](/doc/command-reference/pull) like regular project data (e.g. in
```

  > [Importing and updating fixed revisions](#example-importing-and-updating-fixed-revisions)
  > example below).

- `-s`, `--store` - [enables remote storage](#enabling-remote-storage) for the

>   import, so it can be pushed (e.g. in case the original source is lost).

Reviewer: ```suggestion
  import, so it can be pushed and pulled (e.g. in case the original source is
  lost).
```

> - `-s`, `--store` - turns [store mode](#store-mode) on for the import, so it can
>   be pushed to remote storage (e.g. in case the original source is lost).

Reviewer: ```suggestion
- `-s`, `--store` - [enables remote storage](#enabling-remote-storage) on for
  the import, so it can be pushed (e.g. in case the original source is lost).
```

> - `-s`, `--store` - [enables remote storage](#enabling-remote-storage) on for
>   the import, so it can be pushed (e.g. in case the original source is lost).

Reviewer: ```suggestion
- `-s`, `--store` - [enables remote storage](#enabling-remote-storage) for the
  import, so it can be pushed (e.g. in case the original source is lost).
```
